client/components/data/query-sites/index.jsx

```diff
@@ -3,19 +3,31 @@
  */
 import { Component, PropTypes } from 'react';
 import { connect } from 'react-redux';
-import { bindActionCreators } from 'redux';
 
 /**
  * Internal dependencies
  */
-import { isRequestingSites } from 'state/sites/selectors';
-import { requestSites } from 'state/sites/actions';
+import { isRequestingSites, isRequestingSite } from 'state/sites/selectors';
+import { requestSites, requestSite } from 'state/sites/actions';
 
 class QuerySites extends Component {
-
 	componentWillMount() {
-		if ( ! this.props.requestingSites ) {
-			this.props.requestSites();
+		this.request( this.props );
+	}
+
+	componentWillReceiveProps( nextProps ) {
+		if ( nextProps.siteId !== this.props.siteId ) {
+			this.request( nextProps );
+		}
+	}
+
+	request( props ) {
+		if ( ! props.siteId && ! props.requestingSites ) {
```

Comment:
Fair enough, and if we ever end up paging the results this makes sense too. I think an extra prop works better than giving extra meaning to `null`. Usage then looks like:
```javascript
<QuerySites allSites />
<QuerySites siteId={ siteId } />
```